package/boot/uboot-envtools/files/ramips

```diff
@@ -39,6 +39,9 @@ xiaomi,mir3g|\
 xiaomi,redmi-router-ac2100)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x20000"
 	;;
+hootoo,tm05)
+	ubootenv_add_uci_config "/dev/mtd1" "0x2000" "0x4000" "0x1000"
```

Comment:
Does `fw_printenv` return a valid env list? I'm startled by that 0x2000 start offset is used here.

Reply:
Me too ... and I may have this wrong, I admit. I searched through that partition, and this is where I find information, but not sure it's for u-boot 😞. When I use the more "stock" values though, I get this,

`
Warning: Bad CRC, using default environment
bootcmd=bootp; setenv bootargs root=/dev/nfs nfsroot=${serverip}:${rootpath} ip=${ipaddr}:${serverip}:${gatewayip}:${netmask}:${hostname}::off; bootm
bootdelay=5
baudrate=115200
`

And I know that's not correct (CRC error, wrong baud rate). Thoughts?

Thanks!

Reply:
`hexdump -C /dev/mtd1` might reveal where the envs are stored.

Reply:
Agreed! That's actually what I did - output below,
`
00000000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00002000  b5 2b 34 45 43 6f 75 6e  74 72 79 52 65 67 69 6f  |.+4ECountryRegio|
00002010  6e 3d 30 00 43 6f 75 6e  74 72 79 52 65 67 69 6f  |n=0.CountryRegio|
00002020  6e 41 42 61 6e 64 3d 30  00 43 6f 75 6e 74 72 79  |nABand=0.Country|
00002030  43 6f 64 65 3d 55 53 00  42 73 73 69 64 4e 75 6d  |Code=US.BssidNum|
`

Data from there, to, 
`
00002c10  3d 31 00 73 73 69 64 31  5f 6f 6e 3d 31 00 46 69  |=1.ssid1_on=1.Fi|
00002c20  72 65 57 61 6c 6c 5f 65  6e 3d 30 00 00 00 00 00  |reWall_en=0.....|
00002c30  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00006000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00010000
`

Reply:
OK, that formatted like crap 😞

Reply:
Found it! But this seems odd ... it's in /dev/mtd0. Make any sense?

Here is the info,
0001d200  62 6f 6f 74 63 6d 64 3d  74 66 74 70 00 62 6f 6f  |bootcmd=tftp.boo|
0001d210  74 64 65 6c 61 79 3d 32  00 62 61 75 64 72 61 74  |tdelay=2.baudrat|
0001d220  65 3d 35 37 36 30 30 00  65 74 68 61 64 64 72 3d  |e=57600.ethaddr=|
0001d230  22 30 30 3a 41 41 3a 42  42 3a 43 43 3a 44 44 3a  |"00:AA:BB:CC:DD:|
0001d240  31 30 22 00 69 70 61 64  64 72 3d 31 30 2e 31 30  |10".ipaddr=10.10|
0001d250  2e 31 30 2e 31 32 38 00  73 65 72 76 65 72 69 70  |.10.128.serverip|
0001d260  3d 31 30 2e 31 30 2e 31  30 2e 32 35 34 00 67 61  |=10.10.10.254.ga|
0001d270  74 65 77 61 79 69 70 3d  31 30 2e 31 30 2e 31 30  |tewayip=10.10.10|
0001d280  2e 31 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |.1..............|
0001d290  00 40 1d 00 00 00 00 00  00 00 00 00 00 00 00 00  |.@..............|

Reply:
Maybe, but it might be just default values the bootloader use internally. Perhaps the uboot env is stored in mtd1 0x0 offset with 0x1000 size but it was never used. I would not add the uboot-env config in this case.

Reply:
Well, you gave me an idea - I figured I'd go in to u-boot, set an environment variable, then try to find it. Seemed logical at the time ... LOL. But the result is very odd, to say the least.

So I added a variable (setenv foo This is an example), could see it with printenv (in uboot still), then saved it ... the output from saveenv below,
```
MT7620 # saveenv
Saving Environment to SPI Flash...
raspi_read: from:1d0000 len:10000
Erasing SPI Flash...
raspi_erase: offs:1d0000 len:10000
.
Writing to SPI Flash...
raspi_write: to:1d0000 len:10000
.
done
```

Address is odd 1d0000, looks to be where rootfs is. So then tried to boot, and ...
```
OpenWrt kernel loader for MIPS based SoC
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Looking for OpenWrt image... found at 0xbc051000
Decompressing kernel... failed, data error!

System halted!
```

So saving an env variable in uboot trashed my rootfs ... huh?!?! Doesn't makes sense to me. So I re-flashed (factory.bin, TFTP) ... reboots fine, and the env variable I set is removed. But is it really the case that uboot saveenv writes to where the rootfs is?

Reply:
Try this
from u-boot
check current environment with `printenv`
save it to flash with `saveenv`
see if it shows up in mtd1 with `md 0x30000 100`

Reply:
I think this is what I did just above (meaningless variable, foo) => saved ... and it trashed my rootfs. Or am I missing something?

Reply:
Hmm, that's really odd. Maybe you can verify it by tftp-booting initramfs image and checking the flash. Anyway, it seems it's not possible to save uboot env in a safe place even with the bootloader. I'd say the feature is broken in a sense.

Reply:
BTW you can make a code paragraph by pasting it between lines that start with ``` on each side. Both have to be on a new line

It didn't format well because you did a code line, aka 1 of ` on each side

Reply:
Are we all thinking this is broken (at a basic uboot level), remove it?

Reply:
Yeah...

just wondering though, did it say where it was saving environment to? Could you find it on flash before you tried to boot? (look for it using u-boot)

Reply:
Looks like you had to define a few OEM partitions in the DTS!

Aaaaaand thats would create a hole between kernel and rootfs so you really had to follow that wicked ENS202 and Pisen partitioning scheme ...

Reply:
At least `params` had to be recreated in DTS:

```
$ cat /proc/mtd dev: size erasesize name mtd0: 00800000 00010000 "ALL" mtd1: 00030000 00010000 "Bootloader" mtd2: 00010000 00010000 "Config" mtd3: 00010000 00010000 "Factory" mtd4: 00180000 00010000 "Kernel_RootFS" mtd5: 00010000 00010000 "params" mtd6: 00010000 00010000 "user_backup" mtd7: 00010000 00010000 "user" mtd8: 00600000 00010000 "Rootfs"
```

See OEM Flash layout!
OpenWrt Project: HooToo HT-TM05
https://openwrt.org/toh/hootoo/hootoo_ht-tm05

Reply:
Good catch @xabolcs! That makes sense now 😄. So I can see two options,
1) I can add fixed partitions there, in the DTS, similar to the loader - and then this space will just be skipped / left alone by Openwrt, but uboot can store there.
2) Remove uboot env information, just don't plan to add uboot info (and don't use uboot to save info, or it will mess up Openwrt!)

I'm leaning towards 2), as anything that was there is now gone (after first Openwrt install), and that's a lot of wasted space - the kernel doesn't fit in 1.5 MB, it would have to go in (OEM) mtd8, and (OEM) mtd4 would just sit there unused.

So just get rid of uboot env, agreed?

Reply:
Go for 1) you only need to follow @mpratt14's journey at #3014! :grinning: 

Please re-read carefully [his first comment](https://forum.openwrt.org/t/hootoo-tm-05-add-lzma-loader/67687/25?u=xabolcs)! Mind the `fakeroot` partition ... it's the same as `params` here.


We only need to "waste" that 64K for `u-boot-env`, and that 4K for `loader` partitions. 
Should we follow the talk in our support topic? :wink:

Reply:
Actually (and I may be wrong here!), but I think we waste more like 1.5 MB ... no? That 1.5 MB partition in the middle becomes unusable then, no? That's why I say it's not worth it - throwing away 1.5 MB of an 8 MB flash. But I may be wrong!

I say this because the u-boot env is above that 1.5 MB, so we can't load kernel and flash until even above that (if we try to keep the u-boot env at 0x1D0000).

Thanks!

Reply:
Just read the description of #2299!
Let me quote from it:
> This introduced another problem: We are now wasting the entire 2MB kernel partition.

Reply:
OK, will take a look - thanks. I admit, it does have me wondering ... seems like a lot of work just to enable the u-boot env (as that's all we gain, right?) 😉.

I'm kinda wondering if time is better spent getting the Ravpower device working also => loader is not finding the flash, seems to not be in the same place, but now no serial port 😞.

So I guess it really comes down to - is this worth the pain (right now), or don't make this a blocking issue, get back to it after if really worth it? Thoughts?

Thanks!

Reply:
As @xabolcs is saying, it is possible to save all that space with implementing the virtual_flash and adjusting the partitions and adding padding somehow...but this is very complex and probably not worth it just to have u-boot env

Reply:
Yep, we're saying the same thing, just crossing updates. I'll mark this as resolved - by all means undo that if you disagree! But I believe we're on the same page ... a lot of grief just for u-boot env. I do still plan to read that link though, I am interested.

Thanks!

Reply:
@arrmo , yes the only gain is a correct `u-boot-env` partition, and a safety belt when somebody plays in u-boot with `saveenv`!

Sure, not a blocker here, IMHO, but please mention it in the commit (done already) and put a warning [around the "Flash Layout" on OpenWrt device page](https://openwrt.org/toh/hootoo/hootoo_ht-tm05#flash_layout) about this restriction, with nice steps to reproduce!

Reply:
Yes, that's a very good idea - agreed! I'll add it to the device page, as you mention.. Thanks!

Reply:
Updated the page, added the note - by all means yell if it's not clear!